{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "239 255 223 245 "
     ]
    }
   ],
   "source": [
    "# EX 1: DNA - \tCounting DNA Nucleotides\n",
    "\n",
    "\"\"\"\"\"\n",
    "Given: A DNA string s of length at most 1000 nt.\n",
    "Return: Four integers (separated by spaces) counting the respective number of times that the symbols 'A', 'C', 'G', and 'T' occur in s.\n",
    "\"\"\"\"\"\n",
    "\n",
    "def countNucleotides(input_dna):\n",
    "    \n",
    "    # I create a list within a variable that'll store the counter of the recurrence of each nucleotide\n",
    "    result = [0, 0, 0, 0]\n",
    "    \n",
    "    # I use a for loop to count the recurrence of each nucleotide\n",
    "    for nucleotide in input_dna:\n",
    "        # if nucleotide is 'A' --> increase position 0 in the list \"result\" by 1\n",
    "        if nucleotide ==\"A\":\n",
    "            result[0] = result[0]+1\n",
    "        elif nucleotide ==\"C\":\n",
    "            result[1] = result[1]+1\n",
    "        elif nucleotide ==\"G\":\n",
    "            result[2] = result[2]+1\n",
    "        elif nucleotide ==\"T\":\n",
    "            result[3] = result[3]+1\n",
    "\n",
    "    return result\n",
    "\n",
    "\n",
    "# Statement for test code\n",
    "if __name__ == \"__main__\":\n",
    "    \n",
    "    file = open(\"./datasets-pt1/ex1/rosalind_dna.txt\", \"r\")\n",
    "    rosalind = file.read()\n",
    "\n",
    "    result2 = countNucleotides(rosalind)\n",
    "    for count in result2:\n",
    "        print(count, end=\" \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GGGACUCCCUUAGGCUCAGUCUCAUAUGGUGUUUUCGACUGGGAGUUCAGCCUCGUGUCGAUGAUAGACUAUUGGAAGGGUCGUAGCGCACAUCACAGAUUCCCCAGCAACGUUGCGCUGGGCAGGACUAGAUCGAAAUUGACACAGCCUCCAUCGAUCAUAUCGGUCAGAAGGCCGUUGGCUAGGCAAUGUAGGUUCGAUUUGCUGGGGCGCGUCAUUGAGACCUAAAGAGCUUUAAGUGCCAUCGAUAUAGCCUGAUGGGUGCUGAGGUGUGCUUGGCACCCAGAUGAGGCCACACUCAGCCUGACUCCGCACUUUCCGUUGCUAGAACGGUCGGACGUUUACCGAGGUGUGAAUCACCGACAAAGCAAGGUGUCCUAUAACCGUACAGGAUGUACAAGGAGAAACCAGUGUACGCAUGACAGACAGGCUAGUAUCGAGAGACUCCUCGAUUCUUGUGGAGAGGCAUGACAUUUUGUGCAGCGAUUUUAAAAUCUGCGACUUACUAACGCCGGGGUAUGAUGCGUGUACUCUCGGCAACAGUGGCGGCGUCCAUUCGAUGAAAGAACGGUCCCCUUGGUUCAUCUAAUUAAAGUUAGGGACGGUAGAGAAACCAUCAACCCUGAGCUCUCUUUCAGGAAACUCUCAUACAGUGAUGGGUCAAGAGAGGAGACUUAAAGCGAUUAUAUUCGCGUGACAGUGCCCCGGACAUAUCGUUCAUCCCAGAGGCGUUAGAAGAUAAUUCGACCGGCUUACCCUGCCGUUUAUCAGUACAACACCACCCCUGUAACGUAUAAGCGUGUCGACGCGGAUGUGCGUCCCAAGUGGUAUCGGUUACUUUCCGCUUAAAUAUUGGUGUAGGAAGCACCUCCUCUCCUCUGGCGGUCAGGGGGGGUGUUCACUCCCGUAUCGGUGGUGAGCCUGCCCCAGAUGAGCGGUAAAUAGGUCUAUGCGAGAUCAAACGCGGACUAUCGAUGCGGGGGGUGUUUUUUGGGG\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# EX 2: RNA - Transcribing DNA into RNA\n",
    "\n",
    "\"\"\"\"\"\n",
    "An RNA string is a string formed from the alphabet containing 'A', 'C', 'G', and 'U'.\n",
    "\n",
    "Given a DNA string t corresponding to a coding strand, its transcribed RNA string u is formed by replacing all occurrences of 'T' in t\n",
    "with 'U' in u.\n",
    "\n",
    "Given: A DNA string t having length at most 1000 nt.\n",
    "Return: The transcribed RNA string of t.\n",
    "\"\"\"\"\"\n",
    "\n",
    "def DNAtoRNA(input_dna):\n",
    "    result = ''\n",
    "\n",
    "    # I use the for loop to replace 'T' with 'U' in the DNA string\n",
    "\n",
    "    for nucleotide in input_dna:\n",
    "        # if nucleotide is 'T' --> add 'U' to the string result\n",
    "        if nucleotide ==\"T\":\n",
    "            result +=\"U\"\n",
    "        # otherwise --> add the same nucleotide to the string result\n",
    "        else:\n",
    "            result += nucleotide\n",
    "\n",
    "    return result\n",
    "\n",
    "\n",
    "# Statement for test code\n",
    "if __name__ == \"__main__\":\n",
    "\n",
    "    file = open(\"./datasets-pt1/ex2/rosalind_rna.txt\", \"r\")\n",
    "    rosalind = file.read()\n",
    "\n",
    "    print(DNAtoRNA(rosalind))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GACTATGTTCTGGATGTTGAGTGAGATTTGTGACTCTGATCGGTTATAAGGCTCACTAGTGCTCTAGTCGAAAAGCGTATGCAATCTGACCTGCTGAGGTCAACTGCGGTGTCAGTACGGCTGGGCTGGCTCTACACCTCGCCTTAAACCCCCGAGTACGGGAGTATCGAAATGGCTTGCTCTACAGCCAACCAAGCGTTCACCGATGATCTATATGATACATGAATATGTTTCTCCTGACCTCCGAGTTGATCAAGCACCTGGCTTTAAAAGCATACAAGTGCACCGCTAGTTACTTTCGGCAGTAACTAAGAGTCACTATTTAGCTGGTTCGCTCGTTGAGTTTTCTATTCTCAGTTCCTACCTTGTTGCGGGCGTATGAACCCATACTCTCGTCTGCTCGTTATGGACTGGCGACAGGTTACCTTGTGGGGTACGCGTAGCCAGCACATTACCACGTGAATGGGAACGGCATTGATAAGCCTCCGGGTAGATTTATATTTTGTCCTATGAGGTAATCGACCCAATGTTTGAATATTAAGTTTTAAAACCAGAAAGCCCACCCGAGCTGACCAGGGTTGCGGACCAAATGGTCAGACAGAACAGCGAAGGAAAACAGATCGGTAGGTAACTTAGAGTAGCTAGATGGGACCTGGTTATTTCTTATAGCGACACCACCGGTATGTTCGTCTTAAAGACTTCAAATATTCAATGGAATGGAGGGTTGAAATATCGCCGGGACTAGCCGACCGCTAGCAGCTGAGTAGATGGGAACCTAGATGGTTACCTCGGTCCCTCATGGTACGCTCGCTGTGATCTATAGGTCTGCCCTCAGAGGGACCGAAACTAGGGTCGATAAACGGCTATACTGGGCTGTAAGCCTACTCAAGATTAGTAGCGCTCTGCAACGGAAATTCCAACCATCAAATACGGAACTGTAATGCTATCGAGGTGTTTTTGAAGCATAGAGAG\n"
     ]
    }
   ],
   "source": [
    "# EX 3: REVC - \tComplementing a Strand of DNA\n",
    "\n",
    "\"\"\"\"\"\n",
    "In DNA strings, symbols 'A' and 'T' are complements of each other, as are 'C' and 'G'.\n",
    "\n",
    "The reverse complement of a DNA string s is the string s^c formed by reversing the symbols of s, then taking the complement of \n",
    "each symbol (e.g., the reverse complement of \"GTCA\" is \"TGAC\").\n",
    "\n",
    "Given: A DNA string s of length at most 1000 bp.\n",
    "Return: The reverse complement s^c of s\n",
    "\"\"\"\"\"\n",
    "\n",
    "def ReverseComplement(dna):\n",
    "    \n",
    "    # A dictionary to associate each symbol to its complement\n",
    "    complement = {'A': 'T', 'T': 'A', 'C': 'G', 'G': 'C'}\n",
    "    \n",
    "    # An empty string to contain the reverse complement\n",
    "    reverse_complement = \"\"\n",
    "    \n",
    "    # I use range (range(start, stop, step)) to include the indices of the dna string in reverse order (from last to first). \n",
    "    # With indices you start counting from 0.\n",
    "    # The index of the last character is given by the length of dna - 1 (our \"start\"). \n",
    "    # The arrival limit is -1 (value not included), in order to stop just before index 0 (our \"stop\").\n",
    "    # Our \"step\" is also -1, so that with each iteration the range decreases by one, allowing us to scroll the string backwards.\n",
    "\n",
    "    for i in range(len(dna) - 1, -1, -1):  \n",
    "\n",
    "        # I access each character of the dna string using its index i\n",
    "        base = dna[i]    \n",
    "\n",
    "        # I add the complement of each character to the final string   \n",
    "        reverse_complement += complement[base]  \n",
    "    \n",
    "    return reverse_complement\n",
    "\n",
    "\n",
    "# Statement for test code\n",
    "if __name__ == \"__main__\":\n",
    "\n",
    "    with open(\"./datasets-pt1/ex3/rosalind_revc.txt\", \"r\") as file:\n",
    "        rosalind = file.read().strip()  # strip() deletes unwanted characters before passing the string to the function\n",
    "\n",
    "    print(ReverseComplement(rosalind))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.7588245088245088\n"
     ]
    }
   ],
   "source": [
    "# EX 4: IPRB - \tMendel's First Law\n",
    "\n",
    "\"\"\"\"\"\n",
    "Given: Three positive integers k, m, and n, representing a population containing k+m+n organisms: k individuals are homozygous \n",
    "dominant for a factor, m are heterozygous, and n are homozygous recessive.\n",
    "Return: The probability that two randomly selected mating organisms will produce an individual possessing a dominant allele \n",
    "(and thus displaying the dominant phenotype). Assume that any two organisms can mate.\n",
    "\"\"\"\"\"\n",
    "\n",
    "def dominant_probability(k, m, n):\n",
    "\n",
    "    total = k + m + n\n",
    "\n",
    "    # This formula calculates the total number of possible pairs in a population of total individuals:\n",
    "    # 1. I subtract -1 in view of the fact that the second individual has one less chance compared to the first --> It serves to ensure \n",
    "    # that I am not considering a pair in which the same individual is selected twice.\n",
    "    # 2. I divide by two to avoid counting the same pair twice.\n",
    "     \n",
    "    total_pairs = total * (total - 1) / 2\n",
    "    \n",
    "    # sum of all pairs with dominant offspring, with their specific probabilities\n",
    "    dominant_pairs = (\n",
    "\n",
    "        # AA-AA pairs (100% dominant) - k=AA \n",
    "        # it calculates the number of ways to choose two AA individuals from the population\n",
    "        k * (k - 1) / 2     \n",
    "\n",
    "        # AA-Aa pairs (100% dominant) - m=Aa \n",
    "        # k * m is the number of ways to make a pair between an individual AA and an individual Aa \n",
    "        + k * m\n",
    "\n",
    "        # # AA-aa pairs (100% dominant) - n=aa \n",
    "        # k * n is the number of ways to make a pair between an individual AA and an individual aa             \n",
    "        + k * n\n",
    "\n",
    "        # # Aa-Aa pairs (75% dominant)\n",
    "        # m * (m - 1) / 2 calculates the number of ways of forming pairs between two individuals Aa \n",
    "        # Since these pairs produce dominant offspring only 75% of the time, I multiply this value by 0.75              \n",
    "        + m * (m - 1) / 2 * 0.75 \n",
    "\n",
    "        # Aa-aa pairs (50% dominant)\n",
    "        # m * n is the number of ways to form pairs between individuals Aa and aa\n",
    "        # Since these pairs produce dominant offspring only 50% of the time, I multiply by 0.5\n",
    "        + m * n * 0.5        \n",
    "    )\n",
    "    \n",
    "    # Calculates the overall probability of having a dominant descendant\n",
    "    # dominant_pairs represents the total number of pairs producing dominant offspring\n",
    "    # total_pairs represents all possible pairs.\n",
    "    return dominant_pairs / total_pairs\n",
    "\n",
    "\n",
    "# Statement for test code\n",
    "if __name__ == \"__main__\":\n",
    "\n",
    "    with open(\"./datasets-pt1/ex4/rosalind_iprb.txt\", \"r\") as file:\n",
    "        rosalind = file.read().strip()  # strip() deletes unwanted characters before passing the string to the function\n",
    "\n",
    "        # I need to convert the string read from the file into 3 integers (the ones required by our function - k,m,n):\n",
    "        \n",
    "        # rosalind.split() method splits the string according to spaces and returns a list of values (in this case, strings representing numbers)\n",
    "        # map(int,..) converts each string in the resulting list into an integer\n",
    "        # k, m, n = ... assigns the 3 values in the list to k, m, and n\n",
    "        k, m, n = map(int, rosalind.split())\n",
    "        \n",
    "    print(dominant_probability(k, m, n))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48127306357829\n"
     ]
    }
   ],
   "source": [
    "# EX 5: FIB - Rabbits and Recurrence Relations\n",
    "\n",
    "\"\"\"\"\"\n",
    "Given: Positive integers n≤40 and k≤5\n",
    "Return: The total number of rabbit pairs that will be present after n months, if we begin with 1 pair and in each generation, \n",
    "every pair of reproduction-age rabbits produces a litter of k rabbit pairs (instead of only 1 pair).\n",
    "\"\"\"\"\"\n",
    "\n",
    "def rabbit_pairs(n, k):\n",
    "\n",
    "    # I open a list where the two values represent the number of pairs in the first 2 months (1 pair for each month)\n",
    "    rabbits = [1, 1]\n",
    "    \n",
    "    # A for loop to calculate the number of pairs for each successive month\n",
    "    # I start from month 3 (index 2) and calculate the number of rabbit pairs for each month up to month n\n",
    "    for month in range(2, n):  \n",
    "\n",
    "        # This formula uses the last value in the list (rabbits[-1], representing f(n-1)) \n",
    "        # and the penultimate value (rabbits[-2], representing f(n-2))\n",
    "        # Every month we add the calculated value to the rabbits list with rabbits.append()\n",
    "        rabbits.append(rabbits[-1] + k * rabbits[-2])\n",
    "    \n",
    "    # rabbits[-1] will contain the total number of pairs in the month n\n",
    "    return rabbits[-1]\n",
    "\n",
    "\n",
    "# Statement for test code\n",
    "if __name__ == \"__main__\":\n",
    "\n",
    "    file = open(\"./datasets-pt1/ex5/rosalind_fib.txt\", \"r\")\n",
    "    rosalind = file.read().strip()\n",
    "\n",
    "    n, k = map(int, rosalind.split())\n",
    "\n",
    "    print(rabbit_pairs(n, k))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Rosalind_3915\n",
      "52.866972\n"
     ]
    }
   ],
   "source": [
    "# EX 6: GC - Computing GC Content\n",
    "\n",
    "\"\"\"\"\"\n",
    "Given: At most 10 DNA strings in FASTA format (of length at most 1 kbp each).\n",
    "Return: The ID of the string having the highest GC-content, followed by the GC-content of that string. \n",
    "Rosalind allows for a default error of 0.001 in all decimal answers unless otherwise stated.\n",
    "\"\"\"\"\"\n",
    "\n",
    "def GCcontent(dna):\n",
    "    result = 0\n",
    "\n",
    "    # I create 2 variables to count C and G\n",
    "    count_C = 0\n",
    "    count_G = 0\n",
    "\n",
    "    # every time there will be C or G in the dna string --> add 1 to the counter\n",
    "    for nucleotide in dna:\n",
    "        if nucleotide ==\"C\":\n",
    "            count_C += 1\n",
    "        if nucleotide ==\"G\":\n",
    "            count_G += 1\n",
    "\n",
    "    # I create a new variable that contains the sum of the count of C and G\n",
    "    countCG = count_C + count_G\n",
    "\n",
    "    # I calculate the percentage: count of C and G, divided by the lenght of the dna and multiplied by 100\n",
    "    result = (countCG / len(dna) *100)\n",
    "\n",
    "    return result\n",
    "\n",
    "# This function takes as input a list of strings (fasta_data), which contains data in FASTA format, and returns the ID of the sequence \n",
    "# with the highest GC content, together with the value of that content in percentage.\n",
    "# The fasta_data parameter will be a list of text lines, each representing a line in the FASTA file.\n",
    "def find_highest_GCcontent(fasta_data):\n",
    "\n",
    "    # This variable will store the ID of the sequence with the highest GC content (initially it's an empty string)\n",
    "    highest_gc_id = \"\"\n",
    "\n",
    "    # This variable will store the value of the highest GC content found (initially set to 0.0)\n",
    "    highest_gc_content = 0.0\n",
    "\n",
    "    # This variable stores the ID of the current sequence\n",
    "    current_id = \"\"\n",
    "\n",
    "    # This variable stores the DNA sequence associated with the current ID\n",
    "    # Each time we find a new ID, we reset this variable to build the new sequence\n",
    "    current_sequence = \"\"\n",
    "\n",
    "    # I start a for loop to scroll through each row of fasta_data\n",
    "    for line in fasta_data:\n",
    "\n",
    "        # Removes blank spaces or newline characters (\\n) at the beginning and end of the line\n",
    "        line = line.strip()\n",
    "\n",
    "        # It checks whether the line starts with the character >, which indicates that it is an ID in FASTA format  \n",
    "        if line.startswith(\">\"):  \n",
    "\n",
    "            # This section checks whether we have already constructed a current_sequence (so we are not on the first line of the file)\n",
    "            if current_sequence:\n",
    "\n",
    "                # I call the CGcontent function to calculate the GC content of the current_sequence\n",
    "                gc_content = GCcontent(current_sequence)\n",
    "\n",
    "                # If the GC content of the current sequence is greater than highest_gc_content:\n",
    "                if gc_content > highest_gc_content:\n",
    "\n",
    "                    # I update highest_gc_content with the new GC value\n",
    "                    highest_gc_content = gc_content\n",
    "\n",
    "                    # I update highest_gc_id  with the ID associated with this sequence\n",
    "                    highest_gc_id = current_id\n",
    "\n",
    "            # This part of the code updates the current sequence ID with the new current_id (without the >) \n",
    "            # and resets current_sequence to prepare for collecting the new DNA sequence\n",
    "            current_id = line[1:]\n",
    "            current_sequence = \"\"\n",
    "        \n",
    "        # If the line is not an ID (does not start with >), then it is a part of the DNA sequence\n",
    "        # The += concatenation adds this line to the current_sequence\n",
    "        else:\n",
    "            current_sequence += line  # Aggiunge la sequenza al corrente ID\n",
    "\n",
    "    # This section handles the last sequence found in the file, as it may not have a new > after it to trigger the calculation\n",
    "    if current_sequence:\n",
    "\n",
    "        # I calculate the GC content of current_sequence\n",
    "        gc_content = GCcontent(current_sequence)\n",
    "\n",
    "        # I compare and, if it is greater than highest_gc_content, I update highest_gc_content and highest_gc_id as explained before\n",
    "        if gc_content > highest_gc_content:\n",
    "            highest_gc_content = gc_content\n",
    "            highest_gc_id = current_id\n",
    "\n",
    "    # The function returns the ID of the sequence with the highest GC content and the value of that content.\n",
    "    return highest_gc_id, highest_gc_content\n",
    "\n",
    "# Statement for test code\n",
    "if __name__ == \"__main__\":\n",
    "    with open(\"./datasets-pt1/ex6/rosalind_gc.txt\", \"r\") as file:\n",
    "        fasta_data = file.readlines()\n",
    "    \n",
    "    highest_gc_id, highest_gc_content = find_highest_GCcontent(fasta_data)\n",
    "    print(highest_gc_id)\n",
    "\n",
    "    # It ensures that the value of highest_gc_content is printed with six decimal digits\n",
    "    print(f\"{highest_gc_content:.6f}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MMRACSADPTMLAELVLGRRRHNYSVSRGRLTILIRKGTVVVGCKFHMTLAHRSKCFLFYRHGMHLFGCDRTPPSGSPRGRAQKFGCKDLLSFDTAVRILCMTGLCRLTGHYRLVIVEKEYRASHATYNPLLLHPYFTLPGTLGSSWAWRSDAMLRAPYRLAGSLRTFQGRTFRNSNVARCLFGHTHQLRALATIVCRRCTPFMSGRVPSTLLYLRTRYRSADISNAVLSQPDSSMLIAPLQTHRGTGLDRVYYLHDFGIVIVTYSLFTDKEPFFCIPSVCNRYSVEAILRSVLPGCREAVGAIHYQHRFKHLNALFDTKHQIHAPRSQTENDSVSPITRLSYAPALCLRDSHRSPDTQTGKSAYGLESSVDSVADFLGPNGLERESWNSRQESSSISSRESECTNVTQLNRARKTCLPWPYTVYSSSTLGSNTPHDLPRPVKLPTVGEAASVAASLRFTKQSADCQSESGLWPVVKRKFGTCKHATLKGSTQEVGISTLLATRNVTNGRARAARHKTRTLKVPDGIDKAALHFGQTTHRNSSITRYTNPINLSQDLVHFSINCDITHLREDINKNEKRASSEAVITNGQRATPSGDLCIRSPTRLAPCGIIPLPVASLATGYSDTRSVTTYPLPSHFSVPKPSCRELRDIQICVDAVLSDVVSTSIYYNSQFFIAEDLTSRLCPTWVCTKSKLNYSPSRGTCIVEITGNCPKYGENIHYESLTMVCFHFCANQTRLEFGKPADRTTWGEIGPCPYTIHSSIRRGPRIDPRCCKGRSGLHSGTPASARQLDFDYVWSIDEVSIYGQCTGGRADYTGQDRPGSSYLYSGLHLSRNPATNCQSAPKRSLRDSSQGDRRFRPPFPTAKHFGQPTYDVPQVDYRLHPSYAFTKPPVEVFYPMMISGVGSNLVDDTRFALQYLRSRLLIHDAKSIFTHTILGFASHLVRRSGSKKWTPTRHPLMMHLLHIGNIQLAGSIDAHLQSRATRNRMGGLGSGSAGQHRQRLDHSTISIENSPSDTPQPRGIPSDDCRDPKGDNAHPAFDVCGNQRDEAFRLSWSPKSGTETQTSLSRDHADSTAVQYRREDSRLSCTIVTRRLYFGHVQTGGSLAQSAQSASPMVSHDDTDLVCLTHEGVLESRDVELHHAMPLYKARRLLLRSAAVRGTVEHKRCEVRHPLIVILQFCKSDRHCSTRSYGGSKGKGRTPPARARTCVHRLYACYTGDGSEAPRRREVGSRLSVNSVYLSTGLTCHTPSWYSTVIISKTVHNLWTTPIIFVAIVSTYSYHTTYGTHGVSRVNTCRSVRRFETQIRGPSNGYSHLTCRLDESICDVKSNSGAFRRGLDFNQRVLHQIKNSYSIRELDSQRRYSAPILMWIRVHLFPAHIGSCFINVASRCILTGRHISKNCRKVFADVPQHSVTDISLLNAAAPQLQFGPGYAVSTLSRPFKVGHGIRLLDMHSQVSRSVLPGASGGTHVPGSRAASYVTLKGVKGCRFQHSTDCNTLGSIQVAIAPTGTNESGKAGGRKPNVTDISPSLREGFPFSTTPECRVDALSTRHKQRTPGLGRYVSHDIYRLFHPPLNSLTNHDPLAFRQFGNKWCIRLKDCVDLVKRCPLGPRHHLLACSVETVHSVTRLDFSCLGTPIRWRSNLLVPWQSVVRLVNAVHVISVITNSKRLFCAPGIVMPTYSFLTRCYAGNPGLRATDYFCLTIRGIDFRVVQHMSLMSRTRRSPGVCGSVPNADSPRHKLKTHYINVFPLEGSISIPRIECGRFTTKSYETLLYIVVGRESGQRSMSLLVFTHRLDGPPLSVSRPTLDSLPMNECRIASLIPVLRDMPPRTTVFITGLRIHLDALTMTLVTTVCDHSASSNSMPPAGDRDTCLDAGGPKGLLVSADKRPQNVHLGTNLECAVLSDIIRSDHTSFGTRSLGTGKTERIRATSFWFGLMFGFDSFGLDGRRRRDHMVQRLCILDFLSSPAKDVSWPFESICHLTITSTSPPKSDICWGSNQLRDGLQKHKVTDVVFRARVPDRYLANPTSSWFRRGAECPAYLAGKSHPTTTNETQPYARDSFTRPVTTSTLSLPDSINWVRDFILYIASLFFLARTQRQQPGRSGPLRVGRGSSSCLWQPNYSKTDGMPHRQCIKSVQGYGIDNKDPPPEHDTASRIKDQRKTQLKTVFSLDHSSRSRCSTWLTPLSPRLKIREITLKSNAQSMESKCQVKSVHAVPSSAKYAGYVLRSIPVRMEDSYTTWGRLAHIRESDRRTHPCFPSLHSSSLSTGPSPDRRLNDRAPTSGLSTLGSGPLYATPTEYEKSAGIRLSNGATGEDPLLGGTAKPNQRRNSYEFRNASTVIPEIPQSSARRIHAKPLIGKCGSMVMRAPAGKPDILQANIEKLIPFVELPCPPRPPLIDSGTGFCDRSTSLKSTRRTDARTHSGVSWEHSFRPSIEVEVLDSSIIMGRPHALHVSPRRGSYDIKRSTSQGSAKIRPLIRTTGYRILNEHTGNLVPTTYASPFSSTVRKDGRMAADEARRSPKCPELSLASRTRAQQSPCCDVKLKGSCSALVPVGSNQLNGLVFRTVSPRTCAASVNSGFRVFIQGPVDDSYSTGPQVYSLRELHGPKLPLGSATSNLLDHTEEALDKKGVKYRFGLNLTPVAPGVIGTYSRLSSKTPLGCAMLLLDAAGGCCLWLPSGFDSSSFPLTIVPPSSPTTSFKRLSSAAKGYTRLMVILPARMFHIVLWRAAHQLRRSLHGPIQFNHLLAGGYTIGPSIPPHGATVFGLHPLATRLFPTRNFRCESWVLDEKPCYHPLLSVSLHHCQAECKVYALRVAKNEGRTAVFLLIREPRVNSLIGAEGNALRSVGLTPSITEWTDYLGHLTLVRKDSRRGHAQARYPGLHPRLKHPYLQRYLGDRNDSETHRLKNAGLLDTT\n"
     ]
    }
   ],
   "source": [
    "# EX 7: PROT - Translating RNA into Protein\n",
    "\n",
    "\"\"\"\"\"\n",
    "Given: An RNA string s corresponding to a strand of mRNA (of length at most 10 kbp).\n",
    "Return: The protein string encoded by s.\n",
    "\"\"\"\"\"\n",
    "\n",
    "# The codon table is a map showing how RNA nucleotide sequences (consisting of adenine ‘A’, cytosine ‘C’, guanine ‘G’ and uracil ‘U’) \n",
    "# are translated into amino acids, which are the building blocks of proteins\n",
    "\n",
    "# Each amino acid is encoded by a sequence of 3 nucleotides, called a codon. There are 64 possible combinations of codons \n",
    "# (since 4^3=64), but only 20 common amino acids, so some codons correspond to the same amino acid.\n",
    "# ex: UAA → Stop codon (does not correspond to an amino acid)\n",
    "\n",
    "# Function that takes a string of RNA as input\n",
    "def RNAtoProtein(rna):\n",
    "\n",
    "    # The codon_table variable is a dictionary that maps codons (keys) to amino acids (values)\n",
    "    codon_table = {\n",
    "        'UUU': 'F',    'CUU': 'L',    'AUU': 'I',    'GUU': 'V',\n",
    "        'UUC': 'F',    'CUC': 'L',    'AUC': 'I',    'GUC': 'V',\n",
    "        'UUA': 'L',    'CUA': 'L',    'AUA': 'I',    'GUA': 'V',\n",
    "        'UUG': 'L',    'CUG': 'L',    'AUG': 'M',    'GUG': 'V',\n",
    "        'UCU': 'S',    'CCU': 'P',    'ACU': 'T',    'GCU': 'A',\n",
    "        'UCC': 'S',    'CCC': 'P',    'ACC': 'T',    'GCC': 'A',\n",
    "        'UCA': 'S',    'CCA': 'P',    'ACA': 'T',    'GCA': 'A',\n",
    "        'UCG': 'S',    'CCG': 'P',    'ACG': 'T',    'GCG': 'A',\n",
    "        'UAU': 'Y',    'CAU': 'H',    'AAU': 'N',    'GAU': 'D',\n",
    "        'UAC': 'Y',    'CAC': 'H',    'AAC': 'N',    'GAC': 'D',\n",
    "        'UAA': 'Stop', 'CAA': 'Q',    'AAA': 'K',    'GAA': 'E',\n",
    "        'UAG': 'Stop', 'CAG': 'Q',    'AAG': 'K',    'GAG': 'E',\n",
    "        'UGU': 'C',    'CGU': 'R',    'AGU': 'S',    'GGU': 'G',\n",
    "        'UGC': 'C',    'CGC': 'R',    'AGC': 'S',    'GGC': 'G',\n",
    "        'UGA': 'Stop', 'CGA': 'R',    'AGA': 'R',    'GGA': 'G',\n",
    "        'UGG': 'W',    'CGG': 'R',    'AGG': 'R',    'GGG': 'G'\n",
    "    }\n",
    "\n",
    "    # String to store the translated amino acids.\n",
    "    protein = \"\"\n",
    "\n",
    "    # The loop starts at 0 and continues up to the length of the RNA string, incrementing by 3 at each iteration\n",
    "    # This is used to take each 3-nucleotide sequence (one codon) from the RNA string\n",
    "    for i in range(0, len(rna), 3):\n",
    "       \n",
    "        # This extracts the substring from position i to i + 3\n",
    "        # The notation i:i + 3 indicates that we are taking characters from i up to (but not including) i + 3, \n",
    "        # so we are obtaining a codon consisting of 3 nucleotides\n",
    "        codon = rna[i:i + 3]\n",
    "\n",
    "        # This checks if we have a complete codon of 3 nucleotides\n",
    "        # This is important because at the end of the RNA sequence we may have incomplete nucleotides\n",
    "        if len(codon) == 3:\n",
    "\n",
    "            # Use the codon as a key to search in the codon_table dictionary\n",
    "            # For example, if codon = ‘AUG’, we get amino_acid = ‘M ’ (Methionine)\n",
    "            amino_acid = codon_table[codon]\n",
    "            \n",
    "            # This checks whether we have encountered a stop codon (UAA, UAG, or UGA)\n",
    "            # break immediately exits the for loop, terminating the translation\n",
    "            if amino_acid == 'Stop':\n",
    "                break\n",
    "                \n",
    "            # If it is not a stop codon, it adds the amino acid to the protein string\n",
    "            protein += amino_acid\n",
    "\n",
    "    return protein\n",
    "\n",
    "\n",
    "# Statement for test code\n",
    "if __name__ == \"__main__\":\n",
    "\n",
    "    with open(\"./datasets-pt1/ex7/rosalind_prot.txt\", \"r\") as file:\n",
    "        rosalind = file.read() \n",
    "\n",
    "    print(RNAtoProtein(rosalind))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "50 95 116 169 205 247 267 363 385 428 514 544 552 578 594 601 640 665 690 707 738 763 821 828 874 891 898 "
     ]
    }
   ],
   "source": [
    "# EX 8: SUBS - Finding a Motif in DNA\n",
    "\n",
    "\"\"\"\"\"\n",
    "Given: Two DNA strings s and t (each of length at most 1 kbp).\n",
    "Return: All locations of t as a substring of s.\n",
    "\"\"\"\"\"\n",
    "\n",
    "# I need to find all positions where a shorter string (t) appears within a longer string (s).\n",
    "def find_locations(s, t):\n",
    "\n",
    "    # List for storing the positions found (initially empty).\n",
    "    positions = []\n",
    "\n",
    "    # I calculate the length of the two strings.\n",
    "    len_s = len(s)\n",
    "    len_t = len(t)\n",
    "\n",
    "    # I scroll the string s from left to right, one position at a time.\n",
    "    # The range goes from 0 to len_s - len_t, because the last valid substring of length len_t starts at index len_s - len_t.\n",
    "    for i in range(len_s - len_t + 1):\n",
    "\n",
    "        # In each iteration of the loop, I extract a substring of s of length len_t.\n",
    "        # This substring starts at position i and ends at position i + len_t - 1.\n",
    "        current_substring = s[i:i + len_t]\n",
    "        \n",
    "        # I compare the extracted current_substring with the substring t.\n",
    "        if current_substring == t:\n",
    "\n",
    "            # If we have found a match, we add the position (1-based) to the positions list\n",
    "            # I add 1 because the indexing required by the problem is 1-based, whereas Python uses 0-based indexing.\n",
    "            positions.append(i + 1)\n",
    "    \n",
    "    # At the end of the function, I return the list containing all the positions found.\n",
    "    return positions\n",
    "\n",
    "\n",
    "# Statement for test code\n",
    "if __name__ == \"__main__\":\n",
    "    \n",
    "    with open(\"./datasets-pt1/ex8/rosalind_subs.txt\", \"r\") as file:\n",
    "        \n",
    "        # Since the function requires two arguments (s and t), I must have them both read from the file\n",
    "        s = file.readline().strip()\n",
    "        t = file.readline().strip()\n",
    "\n",
    "    result = find_locations(s, t)\n",
    "    for positions in result:\n",
    "        print(positions, end=\" \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "464\n"
     ]
    }
   ],
   "source": [
    "# EX 9: HAMM - \tCounting Point Mutations\n",
    "\n",
    "\"\"\"\"\"\n",
    "Given two strings s and t of equal length, the Hamming distance between s and t, denoted dH(s,t), is the number of corresponding \n",
    "symbols that differ in s and t.\n",
    "\n",
    "Given: Two DNA strings s and t of equal length (not exceeding 1 kbp).\n",
    "Return: The Hamming distance dH(s,t)\n",
    "\"\"\"\"\"\n",
    "\n",
    "def hamming_distance(s, t):\n",
    "    result = 0\n",
    "\n",
    "    # I use the range function and the positions (i) in order to spot the different \n",
    "    # element (for each position) between the two strings\n",
    "\n",
    "    for i in range(len(s)):\n",
    "\n",
    "        # if the letter in position 'i' in \"s\" is different from the letter \n",
    "        # in position 'i' of \"t\"--> increase the distance (result) by 1\n",
    "         \n",
    "        if s[i] != t[i]:\n",
    "            result += 1\n",
    "\n",
    "    return result\n",
    "\n",
    "\n",
    "\n",
    "# Statement for test code\n",
    "if __name__ == \"__main__\":\n",
    "    \n",
    "    with open(\"./datasets-pt1/ex9/rosalind_hamm.txt\", \"r\") as file:\n",
    "        \n",
    "        # Since the function requires two arguments (s and t), I must have them both read from the file.\n",
    "        s = file.readline().strip()\n",
    "        t = file.readline().strip()\n",
    "\n",
    "    test_result = hamming_distance(s, t)\n",
    "    print (test_result)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cs2",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
